File: lc/reverse-integer.py
#!/usr/bin/env python3
# reverse-integer.py

class Solution:
    def reverse(self, x: int) -> int:
        aux = abs(x)
        res = 0

        while aux > 0:
            res = res * 10 + (aux % 10)
            aux = int(aux / 10)

        if x < 0:
            return -res
        return res

if __name__ == "__main__":

    import unittest

    class TestSolution(unittest.TestCase):

        def helper(self, x: int, want: int) -> None:
            got = Solution.reverse(None, x)
            self.assertEqual(got, want)

        def test_reverse_example_1(self) -> None:
            self.helper(123, 321)

        def test_reverse_example_2(self) -> None:
            self.helper(-123, -321)

        def test_reverse_example_3(self) -> None:
            self.helper(120, 21)

        def test_reverse_example_4(self) -> None:
            self.helper(0, 0)

    unittest.main()

File: lc/reverse-integer.rs
struct Solution;

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let mut aux = match x.checked_abs() {
            Some(v) => v,
            None => return 0,
        };
        let mut res: i32 = 0;

        while aux > 0 {
            res = match res.checked_mul(10) {
                Some(v) => v,
                None => return 0,
            };
            res = match res.checked_add(aux % 10) {
                Some(v) => v,
                None => return 0,
            };
            aux /= 10;
        }

        if x < 0 { return -res; }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn helper(x: i32, want: i32) {
        let got = Solution::reverse(x);
        assert_eq!(got, want);
    }

    #[test]
    fn test_reverse_1() { helper(123, 321); }

    #[test]
    fn test_reverse_2() { helper(-123, -321); }

    #[test]
    fn test_reverse_3() { helper(120, 21); }

    #[test]
    fn test_reverse_4() { helper(0, 0); }

    #[test]
    fn test_reverse_max() { helper(2147483647, 0); }

    #[test]
    fn test_reverse_min() { helper(-2147483648, 0); }

}
